### ISSUE TITLE
Partial blocking when inserting comments in a form

### ISSUE THREAD

**Issue by gizmecano** [CIDRAM/CIDRAM#203]
### CIDRAM version used
`1.17.4-DEV+20267685` (857b40a36a5acdb1607af35703609e056fc16e48)

### PHP version installed
`PHP 5.6.40`

---
## Issue

Quite recently, with an installation of CIDRAM set on an OpenCart store, a rather singular problem appeared.

A visitor has been **partially** blocked: he had the ability to access the main site, but CIDRAM causes an issue as soon as he wanted to add a comment in an order form. Obviously, it seems that one of the CIDRAM templates tried to load when using the form for this particular visitor. See below two screenshots of the error box which were forwarded by the affected visitor.

![error-box-part-1](https://user-images.githubusercontent.com/852661/111279951-8c4d4a00-863b-11eb-8b22-ab72b8e5bfd2.jpg)

![error-box-part-2](https://user-images.githubusercontent.com/852661/111279964-8fe0d100-863b-11eb-8658-f80cf87ea1ea.jpg)

After various checkings, the IP address of the visitor in question is not present in any currently activated block list. To date, this _uncommon_ issue has only occurred with a single visitor and version `1.17` mentioned above: I updated CIDRAM recently and asked this visitor to notify me if the issue persists with the newly released new version `1.18`...

I open this issue (which seems very rare) mainly because I cannot find out in which stratum it is located : trouble due to the IP address, to the CIDRAM hook, to a potential incompatibility between components used by OpenCart and CIDRAM, etc. Any ideas would therefore be welcome.

**Comment by Maikuolan** [comment]
Could be one of the modules causing the problem. Which CIDRAM modules do you have installed?

**Comment by gizmecano** [comment]
> Which CIDRAM modules do you have installed?

At this time, the configuration of this website includes the 3 modules listed below:
  - _Bot Or Browser User Agent Module_ version `2020.154.710`
  - _Optional security extras module_ version `2020.334.685`
  - _Stop Forum Spam module_ version `2021.9.482`

**Comment by Maikuolan** [comment]
Cheers. Investigating now.

**Comment by Maikuolan** [comment]
Not absolutely certain about the cause yet, but I have some ideas.

Can you check whether there are any entries in the logs for the dates in question citing the reason, "`Non-escaped characters in POST`"?

**Comment by gizmecano** [comment]
I just dug through the log files, and the last two entries for this particular visitor don't have the string `Non-escaped characters in POST` you're looking for.

While I don't remember having found records related to the affected IPs when the issue first manifested, I found this time the following lines (I remove the site domain name to only let the requested page):

```log
[Thu, 07 Jan 2021 20:06:11 +0100] "POST /index.php?route=checkout/shipping_method/save HTTP/1.1" 200 3975 "/index.php?route=checkout/checkout" "Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/87.0.4280.88 Safari/537.36"
[Thu, 07 Jan 2021 20:06:50 +0100] "POST /index.php?route=checkout/payment_method/save HTTP/1.1" 200 3973 "/index.php?route=checkout/checkout" "Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/87.0.4280.88 Safari/537.36"
```

**Comment by gizmecano** [comment]
I regret to reopen this issue, but a similar problem has just arisen again. This is only the second time around, but it seems to indicate that it can occur again.

As a reminder, this issue [was initially opened](https://github.com/CIDRAM/CIDRAM/issues/203#issue-832565199) when the version used was `1.17.4-DEV+20267685`. I have since updated to version `1.20.0-DEV+21260503` yesterday without the issue being resolved.

This time, this strange **partial blocking** is explicitly indicated as being due to an `Unexpected token` error (see image 1) and there is not any mention of `Non-escaped characters in POST`. 

![image 1](https://user-images.githubusercontent.com/852661/134645546-8d2ed0e2-1ba6-449f-9e8f-81a00aa784de.png)

Nevertheless, it is CIDRAM which seems to get involved (cf. image 2).

![Image 2](https://user-images.githubusercontent.com/852661/134645587-d018894c-692b-415d-b447-7c3d0aa6fe4d.png)

> Not absolutely certain about the cause yet, but I have some ideas.

I also cannot find any cause of this behavior. The most difficult to understand is this **partial blocking**: if the IP address turns out to be blocked (it is not the case, I checked), the visitor should not have any access to the site in a global way, but not be impacted only for forms...

🤔
 
> Can you check whether there are any entries in the logs for the dates in question citing the reason, "`Non-escaped characters in POST`"?

Very surprisingly, there is no trace of this IP address in any of the log files: I checked the days when the visitor told me he tried to log in and this IP does not appear anywhere.

I think it's important to add a point that could perhaps orientate searching: if the address in question does not appear in any signatures lists, I discovered that it was a [private IP](https://en.wikipedia.org/wiki/Private_network). Maybe there is a setting that I misconfigured or forgot to change for this type of IP address?

**Comment by Maikuolan** [comment]
I think, I'll investigate whether it's possible to detect that the request is a JSON request, and write an optional bypass for such requests against the suspected offending signature. Might do the trick. Will need to test it though. I'll reply back here when I have some more progress. :-)

**Comment by gizmecano** [comment]
Thank you for taking on your time to respond and for thinking about a patch to get around this issue (apparently uncommon).

Do you think that the fact that the address is private has a decisive role in this type of partial blocking?

**Comment by gizmecano** [comment]
> I'll reply back here when I have some more progress.

Just a quick note to see if there is any progress on this (uncommon but intriguing) issue...

**Comment by Maikuolan** [comment]
Not any meaningful progress yet, sorry. (Just some fruitless research, some approaches I attempted, which didn't seem promising, so didn't stick with yet). I'll keep you posted when there's some more progress.

**Comment by gizmecano** [comment]
Since I haven't received any more reports about potential "false positives" related to the issue described above, I guess I can close this now.

**Comment by Maikuolan** [comment]
*For others that might read back over this issue in the future: More information about the resolution and changes implemented can be found at #285.*
